### ISSUE TITLE
Catch wrong input in format_binaries_and_datas()

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#1885]
One of the reasons for #1879 is [format_binaries_and_datas not catching wrong input data](https://github.com/pyinstaller/pyinstaller/issues/1879#issuecomment-195807282).
We shoukd handle this [here](https://github.com/pyinstaller/pyinstaller/blob/v3.1.1/PyInstaller/building/utils.py#L413-414)

**Comment by leycec** [comment]
I obsequiously agree with the spiritual intent of this commit, but not necessarily its current implementation. Previously working hooks and tests are now broken all over the place. That's _not_ necessarily a bad thing, of course; sometimes you have to crack a few skulls to make the world a better place. <sup>**_Raaaaaar!_**<sup>

But is a simple assertion the right way to go about this? Concerns include:

1. For the moment, a non-fatal warning should probably be printed instead. Before we begin raising fatal exceptions for this, we _really_ need to find and fix all erroneous hooks. Sadly, there appear to be a countably infinite number of these hooks. I have sympathy for the man who walks alone.
1. Once we fix that:
  * Currently, `AssertionError` is raised only when Python optimization (e.g., the `-O` CLI option) is disabled. Instead, we should probably _always_ test this condition and raise explicit human-readable exceptions. This may be overkill, but it feels so right inside:

            # This should be "basestr" under Python 2.7. Whatevas!
            if (not isinstance(src_root_path_or_glob, str) or
                not isinstance(trg_root_dir, str)):
                raise ValueError("Source or target path not a string.")
            if not src_root_path_or_glob or not trg_root_dir:
                raise ValueError("Source or target path empty.")

  * `format_binaries_and_datas()` documentation should be revised.
  * `format_binaries_and_datas()` should be refactored to eliminate the existing test for `if not trg_root_dir`.

## tl;dr

Spiritual correctness: **5** :star:
Implementation correctness: **5** :no_good: _"No good!"_ girls.

**Comment by htgoebel** [comment]
I reverted this change.

**Comment by htgoebel** [comment]
I did not expect too many hooks being implemented wrong.

leycec is right about `-O` and that the assert is removed in this case. I did not think about this.<sup>Your point</sup>

Regarding on what has to be tested:
- Testing for these values to be strings is not enough. #1879 was caused by both string being empty strings.
- If the first parameter is empty, this is plain wrong.
- The second parameter is the destination directory. While an empty path is a correct path, it may be a sign of error. So I would prefer to require the second argument to be a non-empty string. [The manual](http://pythonhosted.org/PyInstaller/#adding-data-files) at the moment does not state this to be required, but all examples there are using `'.'` for the second value.
- "Errors should never pass silently." (The Zen of Python) So we should not just issue a warning but fail. The hook is wrong and needs to be fixes.

So my implementation proposal is:
- Update documentation:
  - At "Adding Data Files": "Use `os.curdir` for the top level directory of the bundled app." and "Starting with PyInstaller 3.3, the second value must not be an empty string, use `os.curdir` instead."
  - At "Hooks global variables", "Starting with Pyinstaller 3.3 ..."
- In `format_binaries_and_datas`:
  - Raise a `RuntimeError('Hook created wrong data. Please fix the hook')` if the first value is empty.
  - If the second value is empty, issue a `DeprecationWarning()`
- In Pyinstaller 3.3: If the second value is empty, raise, too.

Comments?

**Comment by springermac** [comment]
This doesn't quite work in practice. :(

For example:

For a path of `path/to/folder/contents` when trying to copy `folder/` to the dist directory.
`datas = ['path/to/folder', '']` currently yields `folder/contents/` in the dist directory.
`datas = ['path/to/folder', os.curdir]` currently yields `contents/` in the dist directory. Not what we wanted.

To achieve what we want with the current implementation without using `''` one would have to split the path given as the first argument and put the last item as the second argument, i.e., `datas = ['path/to/folder', os.path.split('path/to/folder')[1]]`.

In the current implementation of `format_binaries_and_datas()` the empty string as the second argument in the tuple in checked for and used to determine whether to use the `basename` of the path as the folder in the dist directory or not.
